File: 144. Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
		stack<TreeNode*> s;
		auto cur = root;
		vector<int> ans;
		while(!s.empty() || cur) {
			while(cur) {
				s.push(cur);
				ans.push_back(cur -> val);
				cur = cur -> left;
			}
			if(!s.empty()) {
				cur = s.top();
				s.pop();
				cur = cur -> right;
			}
		}
		return ans;
        
    }
};

File: 144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ans, stack, cur = list(), list(), root
        while cur or stack:
            while cur:
                stack.append(cur)
                ans.append(cur.val)
                cur = cur.left
            if stack:
                cur = stack.pop()
                cur = cur.right
        return ans
            